test/alda/lisp/events/note_test.clj

```diff
@@ -0,0 +1,78 @@
+                                :volume 1.0
+                                :tempo/role :master
+                                :last-offset (AbsoluteOffset. -1)
+                                :reference-pitch 440.0
+                                :id "piano-jYJnW"
+                                :quantization 0.9
+                                :tempo 120
+                                :panning 0.5
+                                :current-marker :start
+                                :time-scaling 1
+                                :stock "midi-acoustic-grand-piano"
+                                :track-volume 0.7874015748031497}}
+                 :markers {:start 0}
+                 :cram-level 0
+                 :global-attributes {}
+                 :tempo/values {0 120}
+                 :nicknames {}
+                 :beats-tally-default nil}
+          event {:event-type :note :letter :c :accidentals []
+                 :beats nil :ms nil :slur? nil}
+          processed-event (note/event-updates score event)
+          result-note (-> processed-event
+                          first
+                          :events
+                          first
+                          :midi-note)]
+      (is (= 48 result-note))))
+  #_(testing "MIDI note outside range should raise exception"
+      (let [score {:chord-mode false
+                   :current-instruments #{"piano-Fckbf"}
+                   :events #{}
+                   :beats-tally nil
+                   :instruments {"piano-Fckbf"
+                                 {:octave 10
+                                  :current-offset (AbsoluteOffset. 0)
+                                  :key-signature {}
+                                  :config {:type :midi :patch 1}
+                                  :transposition 0
+                                  :duration {:beats 1}
+                                  :volume 1.0
+                                  :tempo/role :master
+                                  :last-offset (AbsoluteOffset. -1)
+                                  :reference-pitch 440.0
+                                  :id "piano-Fckbf"
+                                  :quantization 0.9
+                                  :tempo 120
+                                  :panning 0.5
+                                  :current-marker :start
+                                  :time-scaling 1
+                                  :stock "midi-acoustic-grand-piano"
+                                  :track-volume 0.7874015748031497}}
+                   :markers {:start 0}
+                   :cram-level 0
+                   :global-attributes {}
+                   :tempo/values {0 120}
+                   :nicknames {}
+                   :beats-tally-default nil}
+            event {:event-type :note :letter :c :accidentals []
+                   :beats nil :ms nil :slur? nil}]
+        (is (thrown? Exception (note/event-updates score event))))))
```

Comment:
This seems to me like a totally reasonable way to test the sad path where the exception is thrown. A couple notes:

* The note needs to be outside of the acceptable MIDI range in order to produce the exception, so one way to achieve this is to start with a new score, then add an "octave set" event to set the octave to a really high or really low octave, then add a C note.

  Here are a couple examples of how to do this in an Alda REPL:

  ```
  > piano: (octave 10) c
  [27713] ERROR data1 out of range: 132
  > piano: (octave -2) c
  [27713] ERROR data1 out of range: -12
  ```

  I think the corresponding Clojure code for these tests would be something like:

  ```clojure
  (let [score (score (part "piano") (octave -2))
        event (note (pitch :c))]
    (is (thrown? Exception (note/event-updates score event))
  ```

Reply:
* Another thing worth considering is whether it might make sense to test that the exception we have in mind is actually being thrown. I've used `thrown-with-msg` for this in the past. In this case, we could maybe do something like:

  ```clojure
  (is (thrown-with-msg?
        clojure.lang.ExceptionInfo
        #"outside of the allowed range"
        (note/event-updates score event)))
  ```